### ISSUE TITLE
Module for editing HTML using DOM-like API

### ISSUE THREAD

**Issue by daveaglick** [statiqdev/Statiq#160]
Similar to HtmlQuery but to edit HTML using an object model instead of outputting results via metadata. Should expose a simple fluent API from the module (`AddAttribute()`, `AddNode()`, `SetName()`, etc). Should also support more complex behavior by accepting a delegate that operates on the underlying objects from AngleSharp (or whichever library).

**Comment by avishnyakov** [comment]
Nice, looks good so far.

Here are some test cases that we implemented with Excecute() and AngleSharp HtmlParser, and which would be nice to wrap up into new module. The following is a preuso-code aiming to describe the desider behaviour:
- [upd] Incorrect markdown render detection

```
// get all nodes with content containing "###something"
// that's typos in the markdown - there must be a space between ### and actual text
```
- [upd] External images detection

```
// get all images with src start with 'http' and put into metadata
// we crafting offline documentation, so need to know what's up with the external images
```
- Setting up class='img-thumbnail' and class='img-responsive' for all img tags, someting like that:

```
Html("img")
    .AddAttribute("class", "img-thumbnail")
    .AddAttribute("class", "img-responsive");
```
- Replacing blockquotes tag and its content with a new HTML structure

```
Html("blockquote").Each( e => {
 e.replaceWith( "<div class='bs-callout'>" + e.InnerHtml() + "</div>" );
});
```
- Making all internal links lower case, remove *.html extension + all external links "_blank"

```
Html("a").Each( e => {

    if(e.Attr('href').StartsWith("/") {

        var href = e.Attr('href');

        e.Attr('target', string.Empty);

        if(href.EndsWith(".html"))
        href = href.Replace(".html", String.Empty);

         e.Attr('href', href.ToLower());

    } else if(e.Attr('href').StartsWith("http") {
         e.Attr('target', "blank");
    }
});
```

All that is only preudo-code. Would not be nice to re0invent the wheel, guess having jQuery similar syntax/API (as csQuery does) is the way to go. Some kind of generic interface might be exposed, and the implementation could be done via either csQuery or AngleSharp.

The chaining is also important. It should be possible to make several queries against the DOM wihtin one batch, meaning that we would avoid multiple steps in the pipeline, avoid multiple loading HTML document and inmcrease performance. Something like that?

```
Html.Query("a")
             .AddAttr("target", "_blank")
             .AddClass("my-super-link")
        .Query("p")
             .Css("padding:10px");
```

having such a module would be a significat step up for Wyam.

Also, I see interesting possibilities on implementing your own module inheriting that HTML module.
Say, Wyam has that HTML module. We craft our own "HtmlComponent" module inheriting from HTML and adding specific behavior. Why? Here are some cases:
- Injecting disqus comments, here we are:

```
public DisqusModule : WyamHtmlModule {
  // implement disqus - specific methods and logic
  // check if disqus IS included in the HTML page
}

//later in the config
Pipelines.Add("site-md-content",

    ReadFiles("*.md"),
    FrontMatter(Yaml()),

        Markdown(),

       Disqus()
          .SetApiKey("mykey")
          .SetThat()
          .SetThis(),

       WriteFiles(".html")
);.
```
- Injecting google analytics:

```
public GAModule : WyamHtmlModule {
  // implement GA - specific methods and logic
  // check if GA is there, in the HTML
}

//later in the config
Pipelines.Add("site-md-content",

    ReadFiles("*.md"),
    FrontMatter(Yaml()),

        Markdown(),

       Disqus()
          .SetApiKey("mykey")
          .SetThat()
          .SetThis(),

      GA()
         .ApiKey("sdsd")
         .WhatEverElseWeNeed(),

       WriteFiles(".html")
);.
```

So what I am trying to say is that Wyam HTML module provides basic jQuery-like access to DOM. Hence we can implement a particuke module to address a specific problem. Just inherit from Wya,m HTML module and i,mplement your component.

Nothing wrong with storing GA scripts in the master page or layour, or having Disqus thing in the markup too. However having these things in the config (or not having them at all) gives so much power and flexibility - wanna GA? Not a problem. Wanna Disqus? Get the NuGet package (!!!), wanna something else? Alright, here we are. 

How do you see that? Seems Wyam can get much more crazy :)

**Comment by daveaglick** [comment]
Holy cow - the idea of injecting specific pre-canned JavaScript snippets using modules is awesome. I could see writing a number of modules for the most popular JS widgets. Once the more generic HTML DOM manipulation module is done, we'll circle back and see what additional modules should be built on top of it.

**Comment by avishnyakov** [comment]
> the idea of injecting specific pre-canned JavaScript snippets using modules is awesome

Yep. Clean layout / master page, and then "production" and "development" pipelines to inject JS widgets. Indeed HTML module is the key to build on top. 

Glad you like the vision :)

**Comment by avishnyakov** [comment]
Put a few more case into the first comment, will be putting the rest of the cases as they arise. Guess that would provide a starting test cases for the module.
